selene_sdk/samplers/dataloader.py

```diff
@@ -0,0 +1,304 @@
+        self.in_memory = in_memory
+        self.unpackbits = unpackbits
+        self._seq_key = seq_key
+        self._tgt_key = tgt_key
+        self.size = size
+
+    def init(func):
+        # delay initialization to allow multiprocessing
+        def dfunc(self, *args, **kwargs):
+            if not self._initialized:
+                self.db = h5py.File(self.file_path, 'r')
+                self.s_len = self.db['{0}_length'.format(self._seq_key)][()]
+                self.t_len = self.db['{0}_length'.format(self._tgt_key)][()]
+                if self.in_memory:
+                    self.sequences = np.asarray(self.db[self._seq_key])
+                    self.targets = np.asarray(self.db[self._tgt_key])
+                else:
+                    self.sequences = self.db[self._seq_key]
+                    self.targets = self.db[self._tgt_key]
+                self._initialized = True
+            return func(self, *args, **kwargs)
+        return dfunc
+
+    @init
+    def __getitem__(self, index):
+        if isinstance(index, int):
+            index = index % self.sequences.shape[0]
+        sequence = self.sequences[index, :, :]
+        targets = self.targets[index, :]
+        if self.unpackbits:
+            sequence = np.unpackbits(sequence, axis=-2)
+            nulls = np.sum(sequence, axis=-1) == 4
+            sequence = sequence.astype(float)
+            sequence[nulls, :] = 0.25
+            targets = np.unpackbits(targets, axis=-1).astype(
+                float)
+        if sequence.ndim == 3:
+            sequence = sequence[:, :self.s_len, :]
+        else:
+            sequence = sequence[:self.s_len, :]
+        if targets.ndim == 2:
+            targets = targets[:, :self.t_len]
+        else:
+            targets = targets[:self.t_len]
+        return (torch.from_numpy(sequence.astype(np.float32)),
+                torch.from_numpy(targets.astype(np.float32)))
+
+    @init
+    def __len__(self):
+        if self.size is None:
+            self.size = self.sequences.shape[0]
+        return self.size
+
+class H5DataLoader(DataLoader):
+    """
+    H5DataLoader that provides optionally parallel sampling from a HDF5
+    dataset that contains sequences and targets data. `H5DataLoader`
+    can be used with `MultiFileSampler` by passing `SamplerDataLoader` object
+    as `train_sampler`, `validate_sampler` or `test_sampler` when initiating a
+    `MultiFileSampler`.
+
+    Parameters
+    ----------
+    file_path : str
+        The file path of the hdf5 file.
+    size : int or None
+        Default is None. Specify dataset size. This be used to limit the dataset to
+        first N rows. If None, use the sequences array length.
+    in_memory : bool, optional
+        Default is False. If True, load entire dataset into memory.
+    num_workers : int, optional
+        Default is 1. If great than 1, use multiple process to parallelize data
+        sampling.
+    use_subset : int, list of int, or None, optional
```

Comment:
@jzthree what's the difference between use_subset and size?

Reply:
good point - what use_subset can do is a superset of size, so we can get rid of `size`?

Reply:
I have removed `size`.